File: Java/lab20_State/task_2_1 – solution/src/Document.java
public class Document {
    private State state; // ініціалізація якимось станом

    public Document() {
        state = new StateDraft();
    }

    public void setState(State state) {
        this.state = state;
    }

    public void publish(String user){
        state.publish(this, user);
    }

    public void abandon(String user){
        state.abandon(this, user);
    }

    // різні методи роботи з документом, що не
    // стосуються шаблону проектування Стан
    private String text = "";

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }
}

File: Cpp/lab20_State/task_2_1 - solution/src/document.cpp
#include "document.hpp"


Document::Document()
: state(new StateDraft()) {}

Document::~Document()
{
  if (this->state != nullptr)
  {
    delete this->state;
    this->state = nullptr;
  }
}

void Document::setState(State* state)
{
  if (this->state != nullptr)
  {
    delete this->state;
  }
  
  this->state = state;
}

void Document::publish(std::string const& user)
{
  this->state->publish(this, user);
}

void Document::abandon(std::string const& user)
{
  this->state->abandon(this, user);
}

std::string const& Document::getText() const
{
  return this->text;
}

void Document::setText(std::string const& text)
{
  this->text = text;
}
